internal/pkg/agent/application/coordinator/coordinator_state.go

```diff
@@ -0,0 +1,207 @@
+// given values instead of its usual internal state. This is used during upgrades.
+// Coordinator will forward the overridden state to any state subscribers.
+// Callers should follow up with ClearOverrideState as soon as it is safe to do so.
+func (c *Coordinator) SetOverrideState(state agentclient.State, message string) {
+	c.overrideStateChan <- &coordinatorOverrideState{
+		state:   state,
+		message: message,
+	}
+}
+
+// ClearOverrideState clears the override state, reverting to reporting Coordinator's
+// real internal state based on the health of its managers.
+func (c *Coordinator) ClearOverrideState() {
+	c.overrideStateChan <- nil
+}
+
+// setRuntimeManagerError updates the error state for the runtime manager.
+// Called on the main Coordinator goroutine.
+func (c *Coordinator) setRuntimeManagerError(err error) {
+	c.runtimeMgrErr = err
+	c.stateNeedsRefresh = true
+}
+
+// setConfigManagerError updates the error state for the config manager.
+// Called on the main Coordinator goroutine.
+func (c *Coordinator) setConfigManagerError(err error) {
+	c.configMgrErr = err
+	c.stateNeedsRefresh = true
+}
+
+// setConfigManagerActionsError updates the error state for the config manager actions errors.
+// Called on the main Coordinator goroutine.
+func (c *Coordinator) setConfigManagerActionsError(err error) {
+	c.actionsErr = err
+	c.stateNeedsRefresh = true
+}
+
+// setVarsManagerError updates the error state for the variables manager.
+// Called on the main Coordinator goroutine.
+func (c *Coordinator) setVarsManagerError(err error) {
+	c.varsMgrErr = err
+	c.stateNeedsRefresh = true
+}
+
+// setOverrideState is the internal helper to set the override state and
+// set stateNeedsRefresh.
+// Must be called on the main Coordinator goroutine.
+func (c *Coordinator) setOverrideState(overrideState *coordinatorOverrideState) {
+	c.overrideState = overrideState
+	c.stateNeedsRefresh = true
+}
+
+// Forward the current state to the broadcaster and clear the stateNeedsRefresh
+// flag. Must be called on the main Coordinator goroutine.
+func (c *Coordinator) refreshState() {
+	c.stateBroadcaster.InputChan <- c.generateReportableState()
+	c.stateNeedsRefresh = false
+}
+
+// applyComponentState merges a changed component state into the overall
+// Coordinator state and sets stateNeedsRefresh.
+// Must be called on the main Coordinator goroutine.
+func (c *Coordinator) applyComponentState(state runtime.ComponentComponentState) {
+	found := false
+	for i, other := range c.state.Components {
+		if other.Component.ID == state.Component.ID {
+			c.state.Components[i] = state
+			found = true
+			break
+		}
+	}
+	if !found {
+		c.state.Components = append(c.state.Components, state)
+	}
+
+	// In the case that the component has stopped, it is now removed.
+	// Broadcast its stopped state immediately, so subscribers get notified of stopped before removal
+	// TODO: is this separation still needed?
```

Comment:
Yes this is needed. Without this it would be impossible to get the reason a unit was stopped. It would just be removed without having a reason reported in the state change. Think of the case that where another application wants to monitor and report the state changes of the Elastic Agent it would always miss this state and message.

Reply:
👍 I think long run it's worth considering other options (like the persistent record of termination states we discussed elsewhere), but I'm happy to leave that for followup conversations, this change by itself will give us a lot more breathing room to make those choices gradually so I'll remove the TODO from this PR